### ISSUE TITLE
Do not use bundles when extensions.csproj exists

### ISSUE THREAD

**Issue by anthonychu** [microsoft/vscode-azurefunctions#1698]
Currently this extension is [checking for the existence](https://github.com/microsoft/vscode-azurefunctions/blob/9d9a73014c5edb17db523ba907ab017ff005dabb/src/utils/verifyExtensionBundle.ts#L59) of the extensions install command in `task.json` and `settings.json` to determine if bundles should be added to `host.json`.

It's probably more reliable to check for the existence of `extensions.csproj`. For instance, this workflow of installing extensions doesn't add the extensions install command in the vscode settings, so the VS Code extension keeps adding bundles back to host.json:

1. Create JS functions project using VS Code extension
1. `func extensions install -p Microsoft.Azure.WebJobs.Extensions.DurableTask -v 2.0.0`
1. Remove bundles from `host.json` if exists
1. Create a durable orchestration function using VS Code extension (this adds bundles into `host.json`)

/cc @ColbyTresness @soninaren

**Comment by EricJizbaMSFT** [comment]
@anthonychu I agree there's room for improvement here, but I'm not sold on the suggested fix. If "extensions install" isn't in tasks.json or settings.json, we won't run `func extensions install` before debugging or deploying, which means their app could fail to run.

As an alternative, perhaps we prompt users saying basically "We think your project should be using bundles, but you still have an extensions.csproj file". At that point we could give them two options:
1. Configure their project to work with extensions.csproj (This will update `tasks.json` and `settings.json` to ensure "func extensions install" is run before debug and deploy)
1. Move over to bundles

**Comment by anthonychu** [comment]
Agreed that running func extensions install before launching, if needed, is a good idea. Also agreed on prompting the user to add the command if they are using extensions.csproj. Do we need a separate issue to track this?

As for the actual check for whether we should add bundles to host.json, I still think that checking for the existence of extensions.csproj is a more robust way to do it than to check for the command as a substring in the .vscode files.

**Comment by soninaren** [comment]
cc @ColbyTresness , @mattchenderson

**Comment by EricJizbaMSFT** [comment]
> As for the actual check for whether we should add bundles to host.json, I still think that checking for the existence of extensions.csproj is a more robust way to do it than to check for the command as a substring in the .vscode files.

I'm proposing we check both. As mentioned, users will fail to run their app if "extensions install" is not set up in their .vscode folder, so I don't want to check _just_ extensions.csproj.

**Comment by EricJizbaMSFT** [comment]
As mentioned, we don't want to check _just_ "extensions.csproj". However, to come up with a hybrid solution checking both "extensions.csproj" and vs code config, we would need to understand more of the scenario here. We want to move people to bundles as much as possible, in fact they're now included in projects by default (https://github.com/microsoft/vscode-azurefunctions/issues/1740). Can you explain more of your scenario here and why you want to manually install an extension instead of using bundles?

**Comment by anthonychu** [comment]
There are three scenarios I can think of:
- Allow custom extensions or extensions not available in bundles (e.g., [Kafka](https://github.com/Azure/azure-functions-kafka-extension), [RabbitMQ](https://github.com/Azure/azure-functions-rabbitmq-extension))
- Allow users to pick and choose the exact combinations of extension versions to use
- Provide access to new versions of extensions before they are available in bundles (e.g. Durable Functions V2). This is the one I'm most interested in and am blocked by.

If someone opts into extensions.csproj, they need to opt out of extensions for all extensions in their project. Currently, adding a function from most templates in VS Code will add bundles back into host.json and breaks the app.

I still think that ignoring bundles if extensions.csproj exists is a good way to solve this. To me, its existence means the user has opted the route of managing their own extensions and they should be able to do this.

**Comment by EricJizbaMSFT** [comment]
What are your steps for debugging and deploying in those scenarios? Assuming users set up their VS Code config to run "extensions install", they'll never get a bundle added. Is there some reason you're opposed to this? (FWIW, I filed a PR to make it easier to set up that config here: https://github.com/microsoft/vscode-azurefunctions/pull/1861)

**Comment by anthonychu** [comment]
My main concern is that checking for `extensions.csproj` is a better indicator and it doesn't rely on the user setting up their VS Code config properly (some projects don't commit their `.vscode` folder so it might need each developer to set it up correctly). That said, it looks like the PR will help greatly with this and will test it out once it's available. Thanks.

**Comment by davidmrdavid** [comment]
Hi @EricJizbaMSFT, thank you the PR! Just double checking, does your PR (https://github.com/microsoft/vscode-azurefunctions/pull/1861) address the issue only at initialization time or will it allow users to change to `extensions.csproj` mid-development.

A common scenario we see is for a user to scaffold their Functions app using the VSCode plugin, find out at runtime that they're experiencing a bug in Durable Extension, and then they decide to manually install the latest patch via `extensions.csproj`. Is this case supported or will the next VSCode-assisted function creation override `host.json` as well. 

We _constantly_ get customer's struggling to patch their Durable Function Apps because of this as they'd forget that VSCode will override their patches over `host.json` every time they add new functions via the editor.

**Comment by manalotoj** [comment]
Spent a few hours troubleshooting a durable functions issue today that required using Microsoft.Azure.WebJobs.Extensions.DurableTask and not the extensions bundle. Unbeknownst to be, the bundle kept being added back to host.json each time I added a function through VS Code. It would be really helpful to have this resolved.

**Comment by EricJizbaMSFT** [comment]
@manalotoj try running "Azure Functions: Initialize Project for Use with VS Code..." from the [command palette](https://code.visualstudio.com/docs/getstarted/userinterface#_command-palette). Assuming you have an "extensions.csproj" file in your project, it will re-configure to use that instead of bundles (regardless of if your project was already "initialized"). After you do that, you can delete the bundle from "host.json" and it should _not_ be added going forward.

@davidmrdavid "extensions.csproj" is a pretty terrible experience compared to bundles (i.e. it requires Node.js users to install .NET). Do we have any idea why people find bugs in the durable extension so often? And why the latest bundle doesn't have the fixes for those bugs? That's pretty concerning to me if this is a "common scenario"

**Comment by davidmrdavid** [comment]
@EricJizbaMSFT , thanks for the heads up on `Azure Functions: Initialize Project for Use with VSCode...`. That seems like a useful workaround in the meantime!

With respect to your question, sometimes bugs happen 😄 . Durable Functions is a complex framework supporting a myriad of languages and so it is hard to maintain and sometimes regressions occur. I did not mean to say that this is a "common scenario" to regular users, but that users with problems will _commonly_ struggle to patch their issue because the editor will work against them, by overriding their `host.json` repeatedly. In truth, this sometimes gets me as well when developing tests with VSCode.

I really think it would be great for our customers if the editor could facilitate this scenario. Otherwise, it makes it difficult to provide hot patches to customers when they need it most.

**Comment by davidmrdavid** [comment]
Just to be clear, we do not want `extensions.csproj` to be a common way to users interact with Functions. That's what bundles are for, of course. 

It's just that sometimes users will have to break the abstraction to get something done, perhaps because our bundle release timeline is blocked or because they need a hot patch urgently, and so in those scenarios, the current experience only makes it harder for them to apply this patch or to get a new feature.

**Comment by anthonychu** [comment]
I'll echo what @davidmrdavid's comments. Our bundles releases are currently coupled with runtime releases, which we all acknowledge is a problem that needs to be addressed. That said, we've listed a few reasons above why folks might need to use extensions.csproj and they are more than just for patching bugs. We're in the early stages of investigating how we can support custom bundles which could address some scenarios but that will not be a near-term solution.

Reinitializing the project assumes the customer knows that they need to do that. If they don't, they end up with the experience that @manalotoj ran into.

Please reconsider using extensions.csproj to decide whether to use bundles, or prompt the customer to reinitialize VS Code when they add a function or when they run the function app.

**Comment by soninaren** [comment]
@EricJizbaMSFT  Not sure what was actually checked-in after the discussion above. Can you summarize the fix in the PR?

**Comment by EricJizbaMSFT** [comment]
1. Don't add extension bundle during 'Create Function' if 'extensions.csproj' exists
1. When users open a Functions project, check to see if they have "extensions.csproj", but their VS Code config isn't set up to run "func: extensions install" before debug/deploy. If so, show this prompt:
<img width="465" alt="Screen Shot 2020-07-07 at 3 27 30 PM" src="https://user-images.githubusercontent.com/11282622/86850638-7c21ab80-c066-11ea-882c-323a1545fe57.png">
